### PR TITLE
add tensor.to to script

### PR DESCRIPTION
**Pull request by eellison**
Previously it only worked with keyword arguments. Now it is fully compatible.

Fix for: https://github.com/pytorch/pytorch/issues/15478

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -4644,6 +4644,72 @@ def func():
...
+            non_blocking = {non_blocking}
+            return {to_str}
+        ''')
+
+        def s(t, to_str, non_blocking=None, device=None):
+            device = device if device is not None else str(t.device)
+            non_blocking = non_blocking if non_blocking is not None else False
+            code = template.format(to_str=to_str, device=device, non_blocking=non_blocking)
+            scope = {}
+            exec(code, globals(), scope)
```
Comment (driazati) [inline test/test_jit.py]: What is this `exec` doing?

File: torch/csrc/jit/register_prim_ops.cpp [hunk 2]
```diff
@@ -86,6 +87,23 @@ static int64_t floordiv(int64_t a, int64_t b) {
...
 
+// reference function THPVariable_to in python_variable_methods.cpp
+static at::Tensor to_dispatch(at::Tensor self, c10::optional<at::Device> device,
+      c10::optional<at::ScalarType> scalarType, bool non_blocking, bool copy) {
+  if (device && device->is_cuda()) {
+    at::globalContext().lazyInitCUDA();
+  }
+  if (!device && !scalarType && !copy) {
+    return self;
+  } else if (!device) {
```
Comment (driazati) [inline torch/csrc/jit/register_prim_ops.cpp]: Is it possible for both `device` and `scalarType` to be `nullopt`s?

File: torch/csrc/jit/register_prim_ops.cpp [hunk 3]
```diff
@@ -256,6 +274,57 @@ RegisterOperators reg({
             return 0;
           };
         }),
+    Operator(
+        "aten::device(str a) -> Device",
+        [](const Node* node) -> Operation {
+          return [](Stack& stack) {
+            push(stack, c10::Device(pop(stack).toStringRef()));
```
Comment (driazati) [inline torch/csrc/jit/register_prim_ops.cpp]: Why `toStringRef`? `IValue` supports `Device`

File: torch/csrc/jit/register_prim_ops.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: I was just replicating the exiting logic which is: 
"""  
if (!device && !scalarType && !copy) {
    Py_INCREF(self);
    return self;
  } else if (!device) {
    return THPVariable_Wrap(dispatch_to(self_, *scalarType, non_blocking, copy));
  } else if (!scalarType) {
    return THPVariable_Wrap(dispatch_to(self_, *device, non_blocking, copy));
  } else {
    return THPVariable_Wrap(dispatch_to(self_, *device, *scalarType, non_blocking, copy));
  }
""" 

I could explicitly check also check for the case where copy=False and device & scalarType aren't set. That wouldn't really make sense in  user code

**Review by driazati — APPROVED** [review]

File: torch/csrc/jit/register_prim_ops.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: This shouldn't have been added - it already exists. Will delete. Good catch

File: torch/csrc/jit/register_prim_ops.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: It errors in pytorch, so not much point in supporting it in script:
```
>>> a.to(copy=True)
Assertion failed: (initialized()), function operator*, file ../c10/util/Optional.h, line 596.
```

**Pull request closed without merging**